### ISSUE TITLE
bizarre empty string is true and both defined/undef

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11556]
Migrated from [rt.perl.org#96326](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=96326) (status was 'resolved')

Searchable as RT96326$

**Comment by p5pRT** [comment]
### From @mauke

#### Created by @mauke



% perl \-wle 'my $x = \*\{\*STDIN\{IO\}\}; defined $x && $x eq "" or die 1; $x and die 2;'
2 at \-e line 1\.

=> $x is the empty string\, yet true

% perl \-wle 'my $x = \*\{\*STDIN\{IO\}\}; defined $x && $x eq "" or die 1; print "\[$x\]"'
\[\]

=> $x is an ordinary \(defined\) empty string

% perl \-wle 'my $x = \*\{\*STDIN\{IO\}\}; defined $x && $x eq "" or die 1; $x \.= ""'
Use of uninitialized value $x in concatenation \(\.\) or string at \-e line 1\.

=> $x is defined\(\) and undef at the same time

Even better\, \.= normally doesn't warn about an uninitialized LHS\!


I don't know what the intended behavior of \*\{\*STDIN\{IO\}\} is but this is probably not it\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

This perlbug was built using Perl 5.12.1 - Thu Jun  3 20:09:15 CEST 2010
It is being executed now by  Perl 5.14.1 - Wed Jun 22 01:57:37 CEST 2011.

Site configuration information for perl 5.14.1:

Configured by mauke at Wed Jun 22 01:57:37 CEST 2011.

Summary of my perl5 (revision 5 version 14 subversion 1) configuration:
   
  Platform:
    osname=linux, osvers=2.6.37-gentoo-r4, archname=i686-linux
    uname='linux nora 2.6.37-gentoo-r4 #3 preempt thu may 5 20:36:24 cest 2011 i686 amd athlon(tm) 64 processor 3200+ authenticamd gnulinux '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=undef, use64bitall=undef, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.6.0', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.12.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.12.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    SAVEARGV0 - disable magic open in <ARGV>


@INC for perl 5.14.1:
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.1/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.1
    /home/mauke/usr/local/lib/perl5/5.14.1/i686-linux
    /home/mauke/usr/local/lib/perl5/5.14.1
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.0/i686-linux
    /home/mauke/usr/local/lib/perl5/site_perl/5.14.0
    /home/mauke/usr/local/lib/perl5/site_perl
    .


Environment for perl 5.14.1:
    HOME=/home/mauke
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LC_COLLATE=POSIX
    LD_LIBRARY_PATH=/home/mauke/usr/local/lib
    LOGDIR (unset)
    PATH=/home/mauke/usr/local/bin:/usr/local/bin:/usr/bin:/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/4.4.5:/opt/sun-jdk-1.4.2.13/bin:/opt/sun-jdk-1.4.2.13/jre/bin:/opt/sun-jdk-1.4.2.13/jre/javaws:/opt/dmd/bin:/usr/games/bin
    PERL_BADLANG (unset)
    PERL_UNICODE=SAL
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From perl@profvince.com


> % perl \-wle 'my $x = \*\{\*STDIN\{IO\}\}; defined $x&&  $x eq "" or die 1; $x and die 2;'
> 2 at \-e line 1\.

This behaves correctly on 5\.8\.x\. On 5\.10\, it segfaults \- a debugging 
5\.10\.1 gives the following assertion failure :

&nbsp;     Assertion \(\(buffer\)\->sv\_flags & 0x00000400\) failed&#8203;: file "sv\.c"\, 
line 1859 at \-e line 1\.

It starts failing on perl 5\.12\.



Vincent\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Fri Aug 05 02&#8203;:42&#8203;:13 2011\, perl@&#8203;profvince\.com wrote&#8203;:
> 
> > % perl \-wle 'my $x = \*\{\*STDIN\{IO\}\}; defined $x&&  $x eq "" or die 1;
> $x and die 2;'
> > 2 at \-e line 1\.
> 
> This behaves correctly on 5\.8\.x\.

In 5\.8\.x\, defined\(\*$io\} returns false\. For defined to return false on a
typeglob seems wrong to me\.

> On 5\.10\, it segfaults \- a debugging
> 5\.10\.1 gives the following assertion failure :
> 
>      Assertion \(\(buffer\)\->sv\_flags & 0x00000400\) failed&#8203;: file "sv\.c"\,
> line 1859 at \-e line 1\.
> 
> It starts failing on perl 5\.12\.

I don’t know the history\, but I suspect stringification was never taken
into account\, hence a string was never provided\.

So this special glob stringifies as undef\, which is a contradiction in
terms; hence all the bugs\.

I would suggest stringifying it as \*\_\_ANON\_\_ or \*\_\_ANONIO\_\_\. \(There’s
precedent for the latter in op\.c&#8203;:ck\_fun\, but I don’t know how to trigger
it\.\)

**Comment by p5pRT** [comment]
### From @Hugmeir

On Mon\, Aug 22\, 2011 at 6&#8203;:38 PM\, Father Chrysostomos via RT \<
perlbug\-followup@&#8203;perl\.org> wrote&#8203;:

>
> I would suggest stringifying it as \*\_\_ANON\_\_ or \*\_\_ANONIO\_\_\. \(There’s
> precedent for the latter in op\.c&#8203;:ck\_fun\, but I don’t know how to trigger
> it\.\)
>
>
perl \-we 'my $t = \{\}; open $t\->\{fh\}\, ">"\, "some\_file\.txt"; lstat $t\->\{fh\}'

That behavior needs some tests\, I think\.

**Comment by p5pRT** [comment]
### From @cpansprout

Now fixed with commit 52a6327\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
